File: App/server/storage.ts
import type { Run, RunConfig, ChatMessage } from "@shared/schema";
import { randomUUID } from "crypto";

export interface IStorage {
  getRuns(): Promise<Run[]>;
  getRun(id: string): Promise<Run | undefined>;
  createRuns(configs: RunConfig[]): Promise<Run[]>;
  getMessages(): Promise<ChatMessage[]>;
  addMessage(message: ChatMessage): Promise<ChatMessage>;
}

export class MemStorage implements IStorage {
  private runs: Map<string, Run>;
  private messages: ChatMessage[];

  constructor() {
    this.runs = new Map();
    this.messages = [];
  }

  async getRuns(): Promise<Run[]> {
    return Array.from(this.runs.values());
  }

  async getRun(id: string): Promise<Run | undefined> {
    return this.runs.get(id);
  }

  async createRuns(configs: RunConfig[]): Promise<Run[]> {
    const runs = configs.map((config) => ({
      id: `run-${randomUUID()}`,
      status: "pending" as const,
      config,
      created_at: new Date().toISOString(),
    }));

    runs.forEach((run) => this.runs.set(run.id, run));
    return runs;
  }

  async getMessages(): Promise<ChatMessage[]> {
    return this.messages;
  }

  async addMessage(message: ChatMessage): Promise<ChatMessage> {
    this.messages.push(message);
    return message;
  }
}

export const storage = new MemStorage();

File: App/backend/storage.py
"""
storage.py — Data persistence layer (placeholder)

This module should handle:
- Storing run configurations and results
- Querying runs by ID, status, or config
- Saving training logs and metrics
- Managing chat message history
- Optional: PostgreSQL integration for production

Backend Implementation Steps:
1. Choose storage backend (SQLite for dev, PostgreSQL for prod)
2. Define schema matching shared/schemas.py
3. Implement CRUD operations for runs
4. Add message storage for chat history
5. Create indexes for efficient querying
6. Add migration support if using SQL

Example Interface:
    storage = Storage(db_path="trex.db")
    
    # Run operations
    run_id = storage.create_run(config)
    storage.update_run_status(run_id, "running")
    storage.update_run_metrics(run_id, {"val_loss": 0.234})
    run = storage.get_run(run_id)
    runs = storage.list_runs(status="completed")
    
    # Message operations
    storage.add_message(message)
    messages = storage.get_messages(limit=50)

This file is intentionally left incomplete for backend teammates to implement.
"""

# TODO: Implement Storage class
# TODO: Choose and set up database (SQLite/PostgreSQL)
# TODO: Define database schema
# TODO: Implement CRUD operations
# TODO: Add query methods with filtering
# TODO: Optional: Add migrations
